store/store_download.go

```diff
@@ -600,6 +659,84 @@ func downloadImpl(ctx context.Context, name, sha3_384, downloadURL string, user
 	return finalErr
 }
 
+var downloadIcon = downloadIconImpl
+
+// downloadIconImpl writes an http.Request which does not require authentication
+// or a progress.Meter.
+func downloadIconImpl(ctx context.Context, name, downloadURL string, w io.ReadWriteSeeker) error {
+	iconURL, err := url.Parse(downloadURL)
+	if err != nil {
+		return err
+	}
+
+	var finalErr error
+	startTime := time.Now()
+	for attempt := retry.Start(downloadRetryStrategy, nil); attempt.Next(); {
+		httputil.MaybeLogRetryAttempt(iconURL.String(), attempt, startTime)
+
+		if cancelled(ctx) {
+			return fmt.Errorf("the download has been cancelled: %s", ctx.Err())
+		}
+
+		clientOpts := &httputil.ClientOptions{} // don't need any options
+		cli := httputil.NewHTTPClient(clientOpts)
+		var resp *http.Response
+		resp, finalErr = doIconRequest(ctx, cli, iconURL)
+		if cancelled(ctx) {
+			return fmt.Errorf("the download has been cancelled: %s", ctx.Err())
+		}
+		if finalErr != nil {
+			if httputil.ShouldRetryAttempt(attempt, finalErr) {
+				continue
+			}
+			break
+		}
+
+		if httputil.ShouldRetryHttpResponse(attempt, resp) {
+			resp.Body.Close()
+			continue
+		}
+
+		// XXX: we're inside retry loop, so this will be closed only on return.
+		defer resp.Body.Close()
```

Comment:
note that defers will pile up with each attempt and are only executed when downloadIconImpl function exits

OTOH there's only a handful of attempts, on the other maybe you could structure the code like this:
```
errRetry := errors.New("retry")
for attempt := retry.Start(downloadRetryStrategy, nil); attempt.Next(); {
	err := func() {
		resp, err := doIconRequest(ctx, cli, iconURL)
		...
		if err != nil {
			if httputil.ShouldRetryAttempt(attempt, err) {
				return errRetry
			}
			return err
		}
		defer resp.Body.Close()
		
		if httputil.ShouldRetryHttpResponse(attempt, resp) {
			return errRetry
		}
		...
	}()

	if err == errRetry {
		continue
	}
	
	if err == nil {
		logger.Debug("Icon download complete")
	}
	return err
}